### PR TITLE
Redo commented code

### LINKED ISSUE #2: Normal (non-block) strings can count as comments
The line `String allFilesRegex = "/*";` gets flagged as "commented out code".

### LINKED ISSUE #4: Completely Revamp String & Comment Processing to Cover All Corner Cases
## Overview
The AutoGrader tests for commented out code among other things. It works correctly 99% of the time, but there are corner cases where it improperly understands the code via static analysis.

## Discussion
Apparently, a few students have bumped in to these corner cases, as evidenced by #3. Students could fix their code to not flag the style checker, or we could keep adjusting the logic to account for the cases as they arise, but a longer-term approach may be to radically upgrade the internal processing of the files.

This is a discussion about how such a radical change may be implemented, and how it may be better for the code in the end.

## Direction
The current implementation evaluates the code line by line which works great when students behave as we normally expect them to behave. However, if comments and strings (or things that look like comments and strings nested inside each other) occur on the same line or span lines unexpectedly, the system could produce unexpected results.

We can rework the related code as a state machine which directly implements the syntax of Java strings and comments, comparable to a parser. With this guarantee that we are always aware of the accurate state of the current characters (`code`, `comment`, or `string`), we can confidently measure and report stats that will be correct.

The best case implementation would be to directly hook in to the CheckStyle Java parser, assuming it has the tools to detect these kings of things. If they can provide all the syntax checking and language parsing for us, then we can focus only on the stats that we are interested in with a much higher level of confidence that the parsing is being done correctly.

> [!IMPORTANT]
> Better than implementing a Java Parser ourselves, would be to leverage a parser built in to an existing library like CheckStyle. Does such a thing exist? Does it meet our needs?

In the event that such hooks are not available, or do not provide enough context to us to be able to collect the stats we need, a comprehensive state diagram is presented here that would achieve the goal.

**The code will be simpler when we align it with the true definition of the syntax anyways.**

## Proposal
Reimplement the CheckStyle static comment string/comment processing to naturally align with the syntax required by the Java language.

## State Diagram
The following state diagram is intended to represent the entire required functionality for the proposed system. The system is gratefully simple because the Java syntax rules are relatively strict and limited.

### Notes and Introduction
Remember the following concepts from CS 252:
* The input is only accepted if the machine ends in an "accepting" state.
    * In this diagram, that can only be reached by reaching EOF `\z` while in the "code," or either of the comment states
    * Reaching EOF from any other state means the strings/literals comments mean the code is incorrect and won't compile anyways.
* The transitions are labelled with regexes.
* This diagram models a _non-deterministic state machine_.
    * This could be easily converted into a deterministic state machine with more verbose transitions
    * Parsing techniques practices in CS 236 enable us to interpret these kinds of languages
* For our purposes, lines of code can be determined by counting the number of newlines `\n` encountered while in the "code" state.
    * This can be generalized to lines of comments too.
* We do not need to handle single quote characters since Java only allows those to represent individual `chars`.
* Java makes it easy to perform this kind of string/comment analysis since it only allows a small set of syntaxes (those listed here):
    * There is no support for string interpolation
    * Backticks are disallowed
* Regex syntax used:
    *  `(?!negative_assertion)` asserts that "negative_assertion" is _not_ the next item in the sequence
    * `[^klass]` represents any character _not_ in the `klass` character class.
    * `\n` represents the _newline_ character
    * `\z` represents _end of file_
* With a little bit of careful ordering of the transitions, it really is this easy to account for all the possible complexities relating to string and comment parsing.

### State Diagram

image

[View source.](https://www.plantuml.com/plantuml/uml/TOyzJyGm38Pt_meJwGHNgUhU0KwmmiADlIBrmpeYdFGKfmE87q_IA2qVOlJZLuzV7yHtpi-6uLMFX6Gk_Xr7ai7fYzUJ1MYA5ilo5eTf980GzuJggEsPgNpKbl1XCeQiLzW92kV4qynU_msODN3Fq_26H--2QbdrabVFT9wvSz-NJW3WUhV3T1avYeSa5eZTbfB2DLOL2AzJyt8wsJQJ6Amna2IB9Sj_L2Qg2Z2yJZTtLqMrpoIfYYgPqZUo5DuSlTIeb89UyXWbWUmpIfP0aYP__blaeWYSR8F2p_2RQXDSHP3hDKKR7lt8YWt8p_s1x3WR_W80) State Diagram produced with [PlantUML](https://plantuml.com/state-diagram).

## Commentary Requested
What do you think about this proposal? Is this an adjustment that would be good overall? Do you think this will ever be worth prioritizing, or should be just close the issue now?

@ThanGerlek @19mdavenport

### PR DESCRIPTION
**Pull request by 19mdavenport**
At @frozenfrank's request, completely reimplemented the commented out code check using the TreeWalker AST rather than parsing through the file itself.

### CODE REVIEW

**Review by ThanGerlek — APPROVED** [review]
I have to admit I don't know how the AST works, but at least on its own it looks good

File: src/main/java/edu/byu/cs240/checkstyle/readability/CommentedCode.java [hunk 1]
```diff
@@ -1,24 +1,25 @@
...
 
 /**
  * Reports successive lines of commented out code
  *
  * @author Michael Davenport
  */
-public class CommentedCode extends AbstractFileSetCheck {
+public class CommentedCode extends AbstractCheck {
 
     private static final Set<Character> CODE_LINE_END_CHARS = Set.of(';', ',', '{', '}', '(', ')', '/');
```
Comment (frozenfrank) [inline src/main/java/edu/byu/cs240/checkstyle/readability/CommentedCode.java:16]: The set of chars feels a little heavy on the syntax in the event of adding a new symbol.

I took a stab at converting it into a simpler format, but I don't think it turned out any better (Java doesn't have concise ways of inserting data into different kinds of collections. This is my attempt, for the record, but I think we should just leave it as is and press forward:

```java

    /** While within a comment block, commented code is detected based on a line ending in any of these symbols */
    private static final String CODE_LINE_END_CHARS_STR = ";,{}()/";
    private static final Set<Character> CODE_LINE_END_CHARS =
                CODE_LINE_END_CHARS_STR.chars()
                .mapToObj(c -> (char) c)
                .collect(Collectors.toSet());
```

I like having the dense view of all the chars, but I don't love the process required to convert them into the set at eval time.

**Review by frozenfrank — APPROVED** [review]
This is amazingly simple! I'm glad we have the additional test cases to increase our confidence that it works properly. Thanks for making this happen so quickly!
